### DISCUSSION TITLE
Traces: function error message not shown

### DISCUSSION THREAD

**Discussion by shykes** [dagger/dagger#12261] [Help]
## Context

In [Traces](https://dagger.cloud/traces/setup), function calls are shown with information about their status:

- If the function returned an error, it's shown in red
- The error message is displayed in a special box - at least in theory

## Problem

The error message shown in the box is not the one actually returned by the function. Instead, it's always the same generic, and useless, message:

`call function "Foo": process "/runtime" did not complete successfully: exit code: 2`

<img width="1380" alt="Screenshot 2024-07-19 at 2 18 47 AM" src="https://github.com/user-attachments/assets/ec561c39-b0cc-4b3a-9caa-3cdfd1169924">

## Solution

Use the existing "error box" to show the actual error returned by the function, instead of the generic "exit code: 2" error.

**Comment by jedevc** [comment]
@shykes is there an issue in the engine/cli/sdks here? Sorry, I'm not sure.

**Comment by vito** [comment]
This is in core:

https://github.com/dagger/dagger/blob/a218fef96bd6d752a21452507785b90fb2a36e70/core/modfunc.go#L254-L261

Here are two competing proposals to kickstart the conversation:

* Just capture `stderr` and return it as the error message, ungarnished (no wrapping). This would be super easy for SDK authors to implement (Go SDK pattern already matches it), also feels like a sane fallback if something blows up unexpectedly (i.e. `panic()`).
* Add a `Function.returnError` API that just takes a string arg. This would be more explicit than above, but SDK authors would have to implement it. This might be more sane for SDKs like Python that print an entire stack trace.

I started with the first point as a quick hack, but now I'm leaning towards the second option, since you'll always have logs to see more details.

There's further temptation to have something like `Function.returnErrorReport` for linters. Arguably that should instead be a structured return value, but the catch is that if you do a regular return value the call won't have failed, which is the whole point of a linter. So it seems there's some merit to the idea of returning some sort of error structure, but I don't know what it should be. Maybe a problem for another day.
